pyccel/parser/syntactic.py

```diff
@@ -1182,25 +1198,48 @@ def _visit_ListComp(self, stmt):
                 lhs = lhs[0]
             else:
                 raise NotImplementedError("A list comprehension cannot be unpacked")
-
         parent = self._context[-2]
         if isinstance(parent, ast.Call):
             output_type = self._visit(parent.func)
         else:
             output_type = 'list'
-
-        index = PyccelSymbol('_', is_temp=True)
-
-        args = [index]
-        target = IndexedElement(lhs, *args)
-        target = Assign(target, result)
-        assign1 = Assign(index, LiteralInteger(0))
-        assign1.set_current_ast(stmt)
-        target.set_current_ast(stmt)
-        generators[-1].insert2body(target)
-        assign2 = Assign(index, PyccelAdd(index, LiteralInteger(1)))
-        assign2.set_current_ast(stmt)
-        generators[-1].insert2body(assign2)
+        operations = []
+        if output_type != 'list':
+            index = PyccelSymbol('_', is_temp=True)
+            args = [index]
+            target = IndexedElement(lhs, *args)
+            target = Assign(target, result)
+            assign1 = Assign(index, LiteralInteger(0))
+            assign1.set_current_ast(stmt)
+            target.set_current_ast(stmt)
+            operations.append(target)
+            assign2 = Assign(index, PyccelAdd(index, LiteralInteger(1)))
+            assign2.set_current_ast(stmt)
+            operations.append(assign2)
+        else:
+            operations.append(DottedName(lhs, FunctionCall('append', [FunctionCallArgument(result)])))
+
+        conditions = []
+        for generator in generators:
+            if generator.body.body:
+                assert isinstance(generator.body.body[0], If)
+                for ifSection in generator.body.body[0].blocks:
+                    conditions.append(ifSection.condition)
+                generator._body = CodeBlock([]) # pylint: disable=protected-access
```

Comment:
I agree with codacy. This is a code smell. Is there no other way to do this?

This will also cause bugs in the AST tree.

Reply:
If the body is not used after this does it matter if it contains ifs?